### ISSUE TITLE
"Unexpected token at end of expression" when importing `typing.Literal` from library

### ISSUE THREAD

**Issue by Jasha10** [microsoft/pyright#3394]
An unexpected error occurs when using `typing.Literal` imported indirectly.

**To Reproduce**
Here are two files that reproduce the issue:
```python
# lib.py
from typing import Literal
__all__ = ["Literal"]
```
```python
# client.py
from lib import Literal
a_123: Literal["1, 2, 3"]
```
Running pyright on `client.py` results in the following:
```text
$ pyright client.py
...
  /home/rig1/dev/nptyping/client.py:3:16 - error: Unexpected token at end of expression
1 error, 0 warnings, 0 informations
...
```

**Expected behavior**
I wouldn't expect any error

**VS Code extension or command-line**
```text
$ pyright --version
pyright 1.1.241
```
Command-line tool

**Additional context**
This came up when I was playing with the [`nptyping`](https://github.com/ramonhagenaars/nptyping) project. The nptyping module contains an import statement `from typing import Literal as Shape` which resulted in the above issue when my downstream code did `from nptyping import Shape`.

**Comment by erictraut** [comment]
`Literal` is a special form exported from the `typing` module. Pyright needs to apply special-case handling for this form early in its analysis — in the "binder" phase, before type evaluation is possible. This special-case handling requires that you import `Literal` from `typing` or `typing_extensions`. If you import and then re-export it from another library, it breaks this special-case handling. 

It is theoretically possible for us to eliminate this limitation, but it would have a measurable impact on analysis performance, so we're reluctant to do so. We could revisit this decision in the future, but for now we recommend that you import `Literal` directly from `typing` or `typing_extensions` to avoid the error you're seeing.

**Comment by erictraut** [comment]
Thinking about this more, I think I've come up with a way to eliminate this limitation without penalizing performance for everyone. It will only affect performance in the rare case where `typing.Literal` is re-exported from another module.

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.242, which I just published. It will also be included in the next release of pylance.
